src/benchmark/file_io_write_micro_benchmark.cpp

```diff
@@ -7,55 +7,175 @@
+  lseek(fd, from * uint32_t_size, SEEK_SET);
+  if (write(fd, data_to_write_start + from, bytes_to_write) != bytes_to_write) {
     close(fd);
-    Fail("Open error:" + std::strerror(errno));
+    Fail("Write error:" + std::strerror(errno));
   }
+}
+
+void write_data_using_pwrite(const size_t from, const size_t to, int32_t fd, uint32_t* data_to_write_start) {
+  const auto uint32_t_size = ssize_t{sizeof(uint32_t)};
+  const auto bytes_to_write = static_cast<ssize_t>(uint32_t_size * (to - from));
+  if (pwrite(fd, data_to_write_start + from, bytes_to_write, from * uint32_t_size) != bytes_to_write) {
+    close(fd);
+    Fail("Write error:" + std::strerror(errno));
+  }
+}
+
+BENCHMARK_DEFINE_F(FileIOWriteMicroBenchmarkFixture, WRITE_NON_ATOMIC_THREADED)(benchmark::State& state) {
+  auto thread_count = static_cast<uint16_t>(state.range(1));
+
+  // for one thread run sequential implementation to avoid measuring unneccesary thread overhead
+  if (thread_count == 1) {
+    auto fd = int32_t{};
+    if ((fd = open(filename, O_WRONLY)) < 0) {
+      close(fd);
+      Fail("Open error:" + std::strerror(errno));
+    }
+
+    for (auto _ : state) {
+      state.PauseTiming();
+      micro_benchmark_clear_disk_cache();
+      state.ResumeTiming();
+
+      lseek(fd, 0, SEEK_SET);
+      if (write(fd, std::data(data_to_write), NUMBER_OF_BYTES) != NUMBER_OF_BYTES) {
+        close(fd);
+        Fail("Write error:" + std::strerror(errno));
+      }
+
+      state.PauseTiming();
+      sanity_check();
+      state.ResumeTiming();
+    }
+
+    close(fd);
+    return;
+  }
+
+  auto filedescriptors = std::vector<int32_t>(thread_count);
+  for (auto i = size_t{0}; i < thread_count; i++) {
+    auto fd = int32_t{};
+    if ((fd = open(filename, O_WRONLY)) < 0) {
+      close(fd);
+      Fail("Open error:" + std::strerror(errno));
+    }
+    filedescriptors[i] = fd;
+  }
+
+  auto threads = std::vector<std::thread>(thread_count);
+  auto batch_size = static_cast<uint64_t>(std::ceil(static_cast<float>(NUMBER_OF_ELEMENTS) / thread_count));
 
   for (auto _ : state) {
     state.PauseTiming();
     micro_benchmark_clear_disk_cache();
     state.ResumeTiming();
 
-    lseek(fd, 0, SEEK_SET);
-    if (write(fd, std::data(data_to_write), NUMBER_OF_BYTES) != NUMBER_OF_BYTES) {
-      close(fd);
-      Fail("Write error:" + std::strerror(errno));
+    auto* data_to_write_start = std::data(data_to_write);
+
+    for (auto i = size_t{0}; i < thread_count; i++) {
+      auto from = batch_size * i;
+      auto to = from + batch_size;
+      if (to >= NUMBER_OF_ELEMENTS) {
+        to = NUMBER_OF_ELEMENTS;
+      }
+      threads[i] = std::thread(write_data_using_write, from, to, filedescriptors[i], data_to_write_start);
+    }
+
+    for (auto i = size_t{0}; i < thread_count; i++) {
+      //Blocks the current thread until the thread identified by *this finishes its execution
+      threads[i].join();
     }
 
     state.PauseTiming();
+    threads.clear();
```

Comment:
this call might not be necessary, as it only clears the threads vector which will be overwritten each time